programs/mango-v4/tests/cases/test_liq_perps_positive_pnl.rs

```diff
@@ -0,0 +1,416 @@
+    let account_0 = make_account(0).await;
+    let account_1 = make_account(1).await;
+
+    //
+    // SETUP: Borrow some spot on account_0, so we can later make it liquidatable that way
+    // (actually borrowing 1000.5 due to loan origination!)
+    //
+    send_tx(
+        solana,
+        TokenWithdrawInstruction {
+            amount: 1000,
+            allow_borrow: true,
+            account: account_0,
+            owner,
+            token_account: payer_mint_accounts[2],
+            bank_index: 0,
+        },
+    )
+    .await
+    .unwrap();
+
+    //
+    // SETUP: Trade perps between accounts
+    //
+    send_tx(
+        solana,
+        PerpPlaceOrderInstruction {
+            account: account_0,
+            perp_market,
+            owner,
+            side: Side::Bid,
+            price_lots,
+            max_base_lots: 10,
+            max_quote_lots: i64::MAX,
+            reduce_only: false,
+            client_order_id: 0,
+        },
+    )
+    .await
+    .unwrap();
+    send_tx(
+        solana,
+        PerpPlaceOrderInstruction {
+            account: account_1,
+            perp_market,
+            owner,
+            side: Side::Ask,
+            price_lots,
+            max_base_lots: 10,
+            max_quote_lots: i64::MAX,
+            reduce_only: false,
+            client_order_id: 0,
+        },
+    )
+    .await
+    .unwrap();
+    send_tx(
+        solana,
+        PerpConsumeEventsInstruction {
+            perp_market,
+            mango_accounts: vec![account_0, account_1],
+        },
+    )
+    .await
+    .unwrap();
+
+    // after this order exchange it is changed by
+    //   10*10*100*(0.5-1) = -5000 for the long account0
+    //   10*10*100*(1-1.5) = -5000 for the short account1
+    // (100 is base lot size)
+    assert_eq!(
+        account_init_health(solana, account_0).await.round(),
+        (10000.0f64 - 1000.5 * 1.4 - 5000.0).round()
+    );
+    assert_eq!(
+        account_init_health(solana, account_1).await.round(),
+        10000.0 - 5000.0
+    );
+
+    //
+    // SETUP: Change the perp oracle to make perp-based health go positive for account_0
+    // perp base value goes to 10*21*100*0.5, exceeding the negative quote
+    // unweighted perp health is 10*1*100*0.5 = 500
+    // but health doesn't exceed 10k because of the 0 overall weight
+    //
+    set_perp_stub_oracle_price(solana, group, perp_market, &base_token, admin, 21.0).await;
+    assert_eq!(
+        account_init_health(solana, account_0).await.round(),
+        (10000.0f64 - 1000.5 * 1.4).round()
+    );
+
+    //
+    // SETUP: Increase the price of the borrow so account_0 becomes liquidatable
+    //
+    set_bank_stub_oracle_price(solana, group, &borrow_token, admin, 10.0).await;
+    assert_eq!(
+        account_init_health(solana, account_0).await.round(),
+        (10000.0f64 - 10.0 * 1000.5 * 1.4).round()
+    );
+
+    //
+    // TEST: Can't liquidate base if health wouldn't go up: no effect
```

Comment:
Some logging in the ix for this case would be a nice to have

Reply:
I've changed the instruction to emit perp balance logs as well as health logs when it doesn't think it can do anything useful